### DISCUSSION TITLE
Why autoload won't pass args?

### DISCUSSION THREAD

**Discussion by dezza** [akinomyoga/ble.sh#542] [Q&A]
Hello again. I was trying to convert some zsh autoload functions to ble.sh

```console
$ ble-autoload .functions/gpf gpf
$ gpf 1
File-argument ($1) missing
```

So initially I thought it was because it lacked either a function closure (`name() {`). But looking at the source it does seem to do what I expect:
https://github.com/akinomyoga/ble.sh/blob/3b2b4b816ef0bacb4e161cf05f2b11cd09617559/src/util.sh#L5254-L5258

I have my own hack for sourcing fpath functions from zsh, this works.
```sh
# zsh stub null
emulate() {
  :
}

# emulates zsh fpath
fload() {
  eval "$2 () { $(<$1)"$'\n'" }"
}
```
Works here:
```console
$ source .functions.bash.sh 
$ fload .functions/gpf gpf
$ gpf 1 # (function executes)
```

Also.. On second invocation it seems to unset itself. 
```console
$ gpf
File-argument ($1) missing
$ gpf
 gpf: command not found
```
Maybe because `ble-import` fails second time, for whatever reason? https://github.com/akinomyoga/ble.sh/blob/3b2b4b816ef0bacb4e161cf05f2b11cd09617559/src/util.sh#L5255-L5258

**Comment by akinomyoga** [comment]
> ```shell
> 
> # emulates zsh fpath
> fload() {
>   eval "$2 () { $(<$1)"$'\n'" }"
> }
> ```

So, you have *only the function body* directly in the file. This is not the expected usage of `ble-autoload`.

`ble-autoload` assumes that the function is *defined* in the file. In other words, `ble-autoload` behaves like Zsh's `autoload -k`, i.e., the ksh-style autoload.

----

> So initially I thought it was because it lacked either a function closure (`name() {`).

Yes, this is the reason.

> But looking at the source it does seem to do what I expect:

How do you think so? The generated stub is basically like `function gpf { source .functions/gpf && gpf "$@"; }`, where `gpf` is expected to be replaced with the atual implementation by `.functions/gpf`.

> Also.. On second invocation it seems to unset itself.

This is because the loaded file did not define the function `gpf`.

> ```
> $ gpf
> File-argument ($1) missing
> ```

This message is output in the "loading phase", but not by the defined function `gpf`.

> ```
> $ gpf
>  gpf: command not found
> ```
>
> Maybe because `ble-import` fails second time, for whatever reason?

It is also failing for the first time. The message you saw in the first attempt of calling `gpf` is output in the loading phase, but not by the actual function `gpf`.

**Reply by dezza** [reply]
> > So initially I thought it was because it lacked either a function closure (`name() {`).
> 
> 
> 
> Yes, this is the reason.

Ahh ok, I misunderstood.. I see, the stub is just for autoloading, not working anything like my POC zsh autoload compat function `fload()`.

I wish on `ble-autoload` there would be an argument for compat with fpath on zsh (no function signature), this would make transitioning much easier and let you keep the pattern (fpath dir) from zsh in your dotfiles (_you don't have to copy the entire fpath dir and add function def to all in a duplicate directory, which is impractical for git repo. Instead they just need to be portable in syntax_)

I tried doing file substitution `<()` into `ble-autoload` to monkey patch the function signature, but since this is then calling `ble-import` it doesn't find the fd.

**Reply by akinomyoga** [reply]
Why don't you use your `fload`? Or, if you really think it's worth, I can accept in the `contrib` repository a PR that implements Zsh's `autoload` emulation. Or you might want to use [an example implementation of `autoload` provided by the official Bash source](https://git.savannah.gnu.org/cgit/bash.git/tree/examples/functions/autoload.v4?h=devel).

I don't think I would extend `ble-autoload`.

- The main purpose of `ble-autoload` is to manage modules of ble.sh, and it defines how the modules should be organized with `ble.sh`. I'm not going to offer redundant freedom of how the modules should be organized. Also, I actually do *not* think the `fpath` approach is a good system; only one function can be defined in one module, or one needs to put as many symbolic links as the public functions defined in each module.
- In addition, with the current interface of `ble-autoload`, one needs to specify both the filename and the function names, which doesn't suit the `fpath` system. I'm not going to change or break the existing interface of `ble-autoload`. The `fpath` system should rather have a different name than `ble-autoload`.

**Reply by dezza** [reply]
Fair point. I will consider the options.

**Reply by dezza** [reply]
@akinomyoga

I now have:

```sh
# zsh stub null
emulate() {
  :
}

# Set trap if new func context detected
bload_ctx_trap() {
  local original_bash_argv0="$1"
  local func_name="$2"
  local traps=("${@:3}")
  echo "${FUNCNAME[-1]}"

  # shellcheck disable=SC2064,SC2068
  if [[ $func_name != "${FUNCNAME[-1]}" ]]; then
    trap "
    # Check if the BASH_ARGV0 needs to be restored
    if [[ $original_bash_argv0 != $BASH_ARGV0 ]]; then
      BASH_ARGV0='$original_bash_argv0'
    fi
    " "${traps[@]}"
  fi
}

bload_check_syntax() {
  local bload_err bload_ret
  bload_err="$(bash -n "$1" 2>&1)"
  bload_ret=$?

  if ! [[ $bload_ret -eq 0 ]]; then
    printf "bload(%s): %s\n" "$(basename "$1")" "$bload_err"
    return $bload_ret
  fi
}

# fpath compat for bash+ble.sh
bload() {
  local file funcname
  local TRAPS=(RETURN)
  file="$(realpath "$1")"; shift
  funcname="$(basename "$file")"

  # shellcheck disable=SC2027,SC2086,SC2294,SC2068,SC2145,SC1078,SC1079
  builtin eval "$funcname () {
  builtin unset -f $funcname

  $funcname () { 
    bload_check_syntax "$file" || return 1
    bload_ctx_trap "$BASH_ARGV0" "${FUNCNAME[0]}" "${TRAPS[@]}"

    if [[ \${BASH_VERSION:0:1} -ge 5 ]]; then
      BASH_ARGV0=$file
    fi
    # \$(< "$file")
    . "$file"
    trap - "${TRAPS[*]}"
  } &&
    #bload_ctx_trap "$BASH_ARGV0" "${FUNCNAME[0]}" "${TRAPS[@]}"

    $funcname \"\$@\"
    trap - "${TRAPS[*]}"
  }
  "
}
```

Does this integrate ok?

**Reply by akinomyoga** [reply]
I feel redundancy with it, but I think it should work.
